File: archive/broomsticks-ios/Broomsticks/FlyingObject.cpp
#include "FlyingObject.h"

FlyingObject::FlyingObject(int iX, int iY, int maxX, int maxY) {
  //graphics = g;
  x = iX;
  y = iY;
  initX = iX;
  initY = iY;

  vx = 0.0f;
  vy = 0.0f;

  catchable = 0;
  isGoldBall = 0;
  pass = 0;

  accel = 2;
  maxSpeed = 6;
  dive = 1;

  w = 10;
  h = 10;

  minx = 0;
    miny = 0; // was 20;
    maxx = maxX;
    maxy = maxY;
  //maxx = g->getWidth();
  //maxy = g->getHeight()-20;

  lastMoveTime = 0;
}

void FlyingObject::reset() {
  x = initX;
  y = initY;
  vx = 0;
  vy = 0;
}

void FlyingObject::draw() {
  // implemented in derived classes
}

//extern int etime;

void FlyingObject::move(int etime, int now) {
  x += vx * etime/40.0f; // v based on 25 fps (40 ms)
  y += vy * etime/40.0f; 
  if (vy < 2) vy += 0.1 * etime/40.0f;
  bounds();
}

void FlyingObject::bounds() {
  int hit = 0;
  if (x < minx) {
    x = minx;
    vx = -vx;
    hit = 1;
  }
  if (x > maxx-w) {
    x = maxx-w;
    vx = -vx;
    hit = 1;
  }
  if (y < miny) {
    y = miny;
    vy = -vy;
    if (vy == 0)
        vy += 0.1;
    hit = 1;
  }
  if (y > maxy-h /* -10 */) {
      y = maxy-h; //-10;
    vy = 0;
    vx = 0;
    hit = 1;
  }
  if (hit)
    pass = 0;
}

void FlyingObject::left() {
  vx -= accel;
  if (vx < -maxSpeed) vx = -maxSpeed;
}

void FlyingObject::right() {
  vx += accel;
  if (vx > maxSpeed) vx = maxSpeed;
}

void FlyingObject::up() {
  vy -= accel;
  if (vy < -maxSpeed) vy = -maxSpeed;
}

void FlyingObject::down() {
    if (!dive) return;
  vy += accel;
  if (vy > maxSpeed) vy = maxSpeed;
}

File: web/public/2/js/FlyingObject.js
// FlyingObject.js - Base class for all moving entities
// Ported from FlyingObject.cxx - implements delta-time physics

export class FlyingObject {
    constructor(game, x, y) {
        this.game = game;
        this.x = x;
        this.y = y;
        this.initX = x;
        this.initY = y;

        this.vx = 0;
        this.vy = 0;

        this.catchable = false;
        this.isGoldBall = false;
        this.pass = false;

        // Physics config - can be overridden per instance
        this.accel = 2;
        this.maxSpeed = 6;
        this.dive = true; // Allow downward movement

        // Dimensions
        this.w = 10;
        this.h = 10;

        // Bounds (set based on game dimensions)
        this.minx = 0;
        this.miny = 20;
        this.maxx = 640;
        this.maxy = 460; // height - 20

        this.lastMoveTime = 0;
    }

    reset() {
        this.x = this.initX;
        this.y = this.initY;
        this.vx = 0;
        this.vy = 0;
        this.pass = false;
    }

    // Delta-time physics: normalized to 25 fps (40ms base frame time)
    move(etime) {
        const dt = etime / 40; // Normalize to 25fps

        this.x += this.vx * dt;
        this.y += this.vy * dt;

        // Gravity: terminal velocity of 2
        if (this.vy < 2) {
            this.vy += 0.1 * dt;
        }

        this.bounds();
    }

    bounds() {
        let hit = false;

        // Left wall
        if (this.x < this.minx) {
            this.x = this.minx;
            this.vx = -this.vx;
            hit = true;
        }

        // Right wall
        if (this.x > this.maxx - this.w) {
            this.x = this.maxx - this.w;
            this.vx = -this.vx;
            hit = true;
        }

        // Ceiling
        if (this.y < this.miny) {
            this.y = this.miny;
            this.vy = -this.vy;
            if (this.vy === 0) {
                this.vy += 0.1;
            }
            hit = true;
        }

        // Floor
        if (this.y > this.maxy - this.h - 10) {
            this.y = this.maxy - this.h - 10;
            this.vy = 0;
            this.vx = 0;
            hit = true;
        }

        // Reset pass state when hitting a wall
        if (hit) {
            this.pass = false;
        }
    }

    left() {
        this.vx -= this.accel;
        if (this.vx < -this.maxSpeed) {
            this.vx = -this.maxSpeed;
        }
    }

    right() {
        this.vx += this.accel;
        if (this.vx > this.maxSpeed) {
            this.vx = this.maxSpeed;
        }
    }

    up() {
        this.vy -= this.accel;
        if (this.vy < -this.maxSpeed) {
            this.vy = -this.maxSpeed;
        }
    }

    down() {
        if (!this.dive) return;
        this.vy += this.accel;
        if (this.vy > this.maxSpeed) {
            this.vy = this.maxSpeed;
        }
    }

    // Bump player (used when hit by ball or other player)
    bump() {
        this.y = 1000; // Send off screen, will respawn at floor
    }

    setAccel(accel) {
        this.accel = accel;
    }

    setMaxSpeed(maxSpeed) {
        this.maxSpeed = maxSpeed;
    }

    getX() { return this.x; }
    getY() { return this.y; }
    getVX() { return this.vx; }
    getVY() { return this.vy; }
    getW() { return this.w; }
    getH() { return this.h; }

    setX(x) { this.x = x; }
    setY(y) { this.y = y; }
    setVX(vx) { this.vx = vx; }
    setVY(vy) { this.vy = vy; }
    setPass(pass) { this.pass = pass; }
}
